node/src/reactor/main_reactor/control.rs

```
@@ -398,11 +398,13 @@ impl MainReactor {
         effect_builder: EffectBuilder<MainEvent>,
     ) -> Result<Effects<MainEvent>, String> {
         info!("{:?}: committing upgrade", self.state);
-        let previous_block_header = match &self.switch_block_header {
-            None => {
-                return Err("switch_block should be Some".to_string());
-            }
-            Some(header) => header.clone(),
+        let previous_block_header = match self
+            .storage
+            .read_highest_switch_block_headers(1)
+            .map(|headers| headers.into_iter().next())
+        {
+            Ok(Some(header)) => header,
+            _ => return Err("The highest complete block should be a switch block".to_string()),
```

Again, I think we should bubble up any storage error.

---

Done